File: zig/rocksdb.zig
const std = @import("std");
const rocksdb = @import("rocksdb");
const toErrCode = @import("common.zig").toErrCode;
var gpa = std.heap.GeneralPurposeAllocator(.{}){};
const allocator = gpa.allocator();

threadlocal var len: u32 = undefined;
threadlocal var err: i32 = undefined;

//-------------------------------------
// Shared functions
//-------------------------------------

// bun-ffi-z: rocksdb_get_len_ptr () ptr
pub export fn rocksdb_get_len_ptr() u64 {
    return @intFromPtr(&len);
}

// bun-ffi-z: rocksdb_get_err_ptr () ptr
pub export fn rocksdb_get_err_ptr() u64 {
    return @intFromPtr(&err);
}

pub export fn rocksdb_free_(data: u64) void {
    rocksdb.free(@ptrFromInt(data));
}

//-------------------------------------
// DB functions
//-------------------------------------

// bun-ffi-z: rocksdb_db_open (ptr, bool, bool, bool, u32, i32) ptr
pub export fn rocksdb_db_open(
    path: [*c]const u8,
    create_if_missing: bool,
    error_if_exists: bool,
    paranoid_checks: bool,
    write_buffer_size: u32,
    max_open_files: i32,
) u64 {
    var mode: rocksdb.Options.OpenMode = rocksdb.Options.OpenMode.must_exist;

    if (create_if_missing == true) {
        mode = rocksdb.Options.OpenMode.create_if_missing;
    }

    if (error_if_exists == true) {
        mode = rocksdb.Options.OpenMode.exclusive_create;
    }

    const dbp = allocator.create(rocksdb.Database) catch return 0;

    dbp.* = rocksdb.Database.open(std.mem.span(path), rocksdb.Options{
        .mode = mode,
        .paranoid_checks = paranoid_checks,
        .write_buffer_size = write_buffer_size,
        .max_open_files = max_open_files,
    }) catch |e| {
        err = toErrCode(e);
        return 0;
    };

    return @intFromPtr(dbp);
}

pub export fn rocksdb_db_close(db_ptr: u64) void {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);
    db.close();
    allocator.destroy(db);
}

pub export fn rocksdb_db_destroy(path: [*c]const u8) i32 {
    rocksdb.Database.destroy(std.mem.span(path)) catch |e| {
        err = toErrCode(e);
    };
    return 0;
}

pub export fn rocksdb_db_put(db_ptr: u64, key: [*c]const u8, key_len: u32, value: [*c]const u8, value_len: u32) i32 {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);
    db.put(key[0..key_len], value[0..value_len]) catch |e| {
        err = toErrCode(e);
        return 0;
    };
    return 0;
}

// bun-ffi-z: rocksdb_db_get (u64, ptr, u32) ptr
pub export fn rocksdb_db_get(db_ptr: u64, key: [*c]const u8, key_len: u32) u64 {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);

    const value = db.get(key[0..key_len]) catch |e| {
        err = toErrCode(e);
        return 0;
    } orelse {
        err = 0;
        return 0;
    };
    len = @intCast(value.len);
    return @intFromPtr(value.ptr);
}

pub export fn rocksdb_db_delete(db_ptr: u64, key: [*c]const u8, key_len: u32) i32 {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);
    db.delete(key[0..key_len]) catch |e| return toErrCode(e);
    return 0;
}

//-------------------------------------
// Batch functions
//-------------------------------------

// / bun-ffi-z: rocksdb_writebatch_create_ () ptr
pub export fn rocksdb_writebatch_create_() u64 {
    const batch = rocksdb.WriteBatch.init() catch return 0;
    return @intFromPtr(batch.handle);
}

pub export fn rocksdb_writebatch_destroy_(ptr: u64) void {
    var batch = rocksdb.WriteBatch{ .handle = @ptrFromInt(ptr) };
    batch.destroy();
}

pub export fn rocksdb_writebatch_clear_(ptr: u64) void {
    var batch = rocksdb.WriteBatch{ .handle = @ptrFromInt(ptr) };
    batch.clear();
}

pub export fn rocksdb_writebatch_put_(ptr: u64, key: [*c]const u8, key_len: u32, value: [*c]const u8, value_len: u32) void {
    var batch = rocksdb.WriteBatch{ .handle = @ptrFromInt(ptr) };
    batch.put(key[0..key_len], value[0..value_len]);
}

pub export fn rocksdb_writebatch_delete_(ptr: u64, key: [*c]const u8, key_len: u32) void {
    var batch = rocksdb.WriteBatch{ .handle = @ptrFromInt(ptr) };
    batch.delete(key[0..key_len]);
}

pub export fn rocksdb_db_write(db_ptr: u64, batch_ptr: u64) i32 {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);
    var batch = rocksdb.WriteBatch{ .handle = @ptrFromInt(batch_ptr) };

    db.write(&batch) catch |e| return toErrCode(e);
    return 0;
}

//-------------------------------------
// Iterator functions
//-------------------------------------
// bun-ffi-z: rocksdb_db_create_iterator (u64) ptr
pub export fn rocksdb_db_create_iterator(db_ptr: u64) u64 {
    const db: *rocksdb.Database = @ptrFromInt(db_ptr);
    const iter_ptr = allocator.create(rocksdb.Iterator) catch return 0;
    iter_ptr.* = rocksdb.Iterator.init(db, .{}) catch |e| {
        err = toErrCode(e);
        return 0;
    };
    return @intFromPtr(iter_ptr);
}

pub export fn rocksdb_iterator_destroy(iter_ptr: u64) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.destroy();
    allocator.destroy(iter);
}

pub export fn rocksdb_iterator_valid(iter_ptr: u64) bool {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    return iter.valid();
}

pub export fn rocksdb_iterator_seek_to_first(iter_ptr: u64) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.seek_to_first();
}

pub export fn rocksdb_iterator_seek_to_last(iter_ptr: u64) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.seek_to_last();
}

pub export fn rocksdb_iterator_seek(iter_ptr: u64, key: [*c]const u8, key_len: u32) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.seek(key[0..key_len]);
}

pub export fn rocksdb_iterator_next(iter_ptr: u64) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.next();
}

pub export fn rocksdb_iterator_prev(iter_ptr: u64) void {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.prev();
}

// bun-ffi-z: rocksdb_iterator_key (u64) ptr
pub export fn rocksdb_iterator_key(iter_ptr: u64) u64 {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    const key = iter.key();
    len = @intCast(key.len);
    return @intFromPtr(key.ptr);
}

// bun-ffi-z: rocksdb_iterator_value (u64) ptr
pub export fn rocksdb_iterator_value(iter_ptr: u64) u64 {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    const value = iter.value();
    len = @intCast(value.len);
    return @intFromPtr(value.ptr);
}

pub export fn rocksdb_iterator_get_error(iter_ptr: u64) i32 {
    const iter: *rocksdb.Iterator = @ptrFromInt(iter_ptr);
    iter.get_error() catch |e| return toErrCode(e);
    return 0;
}

File: src/rocksdb.ts
import { JSCallback, read, toArrayBuffer, type Pointer } from "bun:ffi";
import { binding } from "./binding.ts";
import { throwErr } from "./common.ts";

// ptrs to hold length and error codes from zig binding
const lenPtr = binding.rocksdb_get_len_ptr() as Pointer;
const errPtr = binding.rocksdb_get_err_ptr() as Pointer;

const textEncoder = new TextEncoder();

export type DB = number & { rocksdb_db: never };

export function dbOpen(
  path: string,
  options: {
    create_if_missing?: boolean;
    error_if_exists?: boolean;
    paranoid_checks?: boolean;
    write_buffer_size?: number;
    max_open_files?: number;
  } = {}
): DB {
  const db = binding.rocksdb_db_open(
    textEncoder.encode(path + "\0"),
    options.create_if_missing ?? false,
    options.error_if_exists ?? false,
    options.paranoid_checks ?? true,
    options.write_buffer_size ?? 4 * 1024 * 1024,
    options.max_open_files ?? 1000
  );
  if (db === null) {
    throwErr(read.i32(errPtr, 0));
  }
  return db as number as DB;
}

export function dbClose(db: DB): void {
  binding.rocksdb_db_close(db);
}

export function dbDestroy(path: string): void {
  throwErr(binding.rocksdb_db_destroy(textEncoder.encode(path + "\0")));
}

export function dbPut(db: DB, key: Uint8Array, value: Uint8Array): void {
  throwErr(binding.rocksdb_db_put(db, key, key.length, value, value.length));
}

export function dbGet(db: DB, key: Uint8Array): Uint8Array | null {
  const valuePtr = binding.rocksdb_db_get(db, key, key.length);
  if (valuePtr === null) {
    throwErr(read.i32(errPtr, 0));
    return null;
  }
  const valueLen = read.u32(lenPtr, 0);
  const value = new Uint8Array(toArrayBuffer(valuePtr, 0, valueLen).slice());
  binding.rocksdb_free_(valuePtr);
  return value;
}

export function dbDelete(db: DB, key: Uint8Array): void {
  throwErr(binding.rocksdb_db_delete(db, key, key.length));
}

export function dbBatchPut(
  db: DB,
  batch: { key: Uint8Array; value: Uint8Array }[]
): void {
  const batchPtr = binding.rocksdb_writebatch_create_();
  for (const { key, value } of batch) {
    binding.rocksdb_writebatch_put_(
      batchPtr,
      key,
      key.length,
      value,
      value.length
    );
  }
  const result = binding.rocksdb_db_write(db, batchPtr);
  binding.rocksdb_writebatch_destroy_(batchPtr);
  throwErr(result);
}

export function dbBatchDelete(db: DB, batch: Uint8Array[]): void {
  const batchPtr = binding.rocksdb_writebatch_create_();
  for (const key of batch) {
    binding.rocksdb_writebatch_delete_(batchPtr, key, key.length);
  }
  const result = binding.rocksdb_db_write(db, batchPtr);
  binding.rocksdb_writebatch_destroy_(batchPtr);
  throwErr(result);
}

export type Iterator = number & { rocksdb_iterator: never };

export function dbIterator(db: DB): Iterator {
  return binding.rocksdb_db_create_iterator(db) as number as Iterator;
}

export function iteratorDestroy(it: Iterator): void {
  binding.rocksdb_iterator_destroy(it);
}

export function iteratorSeekToFirst(it: Iterator): void {
  binding.rocksdb_iterator_seek_to_first(it);
}

export function iteratorSeekToLast(it: Iterator): void {
  binding.rocksdb_iterator_seek_to_last(it);
}

export function iteratorSeek(it: Iterator, key: Uint8Array): void {
  binding.rocksdb_iterator_seek(it, key, key.length);
}

export function iteratorNext(it: Iterator): void {
  binding.rocksdb_iterator_next(it);
}

export function iteratorPrev(it: Iterator): void {
  binding.rocksdb_iterator_prev(it);
}

export function iteratorValid(it: Iterator): boolean {
  return binding.rocksdb_iterator_valid(it);
}

export function iteratorKey(it: Iterator): Uint8Array {
  const keyPtr = binding.rocksdb_iterator_key(it);
  if (keyPtr === null) {
    throwErr(read.i32(errPtr, 0));
  }
  const keyLen = read.u32(lenPtr, 0);
  return new Uint8Array(toArrayBuffer(keyPtr as Pointer, 0, keyLen).slice());
}

export function iteratorValue(it: Iterator): Uint8Array {
  const valuePtr = binding.rocksdb_iterator_value(it);
  if (valuePtr === null) {
    throwErr(read.i32(errPtr, 0));
  }
  const valueLen = read.u32(lenPtr, 0);
  return new Uint8Array(
    toArrayBuffer(valuePtr as Pointer, 0, valueLen).slice()
  );
}

export function iteratorError(it: Iterator): void {
  throwErr(binding.rocksdb_iterator_get_error(it));
}
